Chore: Configure the composer package for publishing

gziolo opened the pull request:
- Closes https://github.com/WordPress/abilities-api/issues/13.

Prepare the composer package for publication and consumption by WordPress plugins.

For complete transparency, I reused some of the configurations from:
- https://github.com/WordPress/php-ai-client/blob/trunk/composer.json
- https://github.com/WordPress/php-ai-client/blob/trunk/.gitattributes
- https://github.com/WordPress/gutenberg/blob/trunk/.gitattributes

I also used AI (GitHub Copilot coding agent) extensively to configure everything so it works both as a WP plugin and shares only `includes` through Composer. I avoided using namespaces to keep things simplified and used `autoload.files` to bootstrap all necessary files.

## Testing

Some checks that I executed

### Validate composer config

```bash
composer validate --no-check-publish
```

Result contained deprecation notices and a success message:

```
./composer.json is valid
```

### Files included in the git archive

```bash
echo "Total tracked files:" && git ls-files | wc -l && echo "\nFiles in git archive (respecting .gitattributes export-ignore):" && git archive --format=tar HEAD | tar -tf - | wc -l && echo "\nFiles excluded from archive (tracked but export-ignored):" && comm -23 <(git ls-files | sort) <(git archive --format=tar HEAD | tar -tf - | sort) | sed -n '1,200p' && echo "\nFiles explicitly in archive (sample):" && git archive --format=tar HEAD | tar -tf - | sed -n '1,200p'
```

Result

```
Total tracked files:
      40

Files in git archive (respecting .gitattributes export-ignore):
      22

Files excluded from archive (tracked but export-ignored):
.editorconfig
.gitattributes
.github/workflows/copilot-setup-steps.yml
.github/workflows/release.yml
.github/workflows/test.yml
.gitignore
.nvmrc
.prettierignore
.prettierrc.js
.wordpress-org/.gitkeep
.wp-env.json
composer.lock
package-lock.json
package.json
phpcs.xml.dist
phpstan.neon.dist
phpunit.xml.dist
tests/_output/.gitkeep
tests/bootstrap.php
tests/unit/abilities-api/wpAbilitiesRegistry.php
tests/unit/abilities-api/wpRegisterAbility.php
tests/unit/rest-api/wpRestAbilitiesInit.php
tests/unit/rest-api/wpRestAbilitiesListController.php
tests/unit/rest-api/wpRestAbilitiesRunController.php

Files explicitly in archive (sample):
.wordpress-org/
CONTRIBUTING.md
LICENSE.md
README.md
abilities-api.php
composer.json
docs/
docs/1.intro.md
docs/2.getting-started.md
docs/3.registering-abilities.md
docs/4.using-abilities.md
includes/
includes/abilities-api.php
includes/abilities-api/
includes/abilities-api/class-wp-abilities-registry.php
includes/abilities-api/class-wp-ability.php
includes/bootstrap.php
includes/rest-api/
includes/rest-api/class-wp-rest-abilities-init.php
includes/rest-api/endpoints/
includes/rest-api/endpoints/class-wp-rest-abilities-list-controller.php
includes/rest-api/endpoints/class-wp-rest-abilities-run-controller.php
```

gziolo commented on abilities-api.php:
```diff
@@ -27,26 +27,17 @@
- */
-require_once WP_ABILITIES_API_DIR . 'includes/rest-api/class-wp-rest-abilities-init.php';
+if ( function_exists( 'add_action' ) ) {
+	add_action( 'rest_api_init', array( 'WP_REST_Abilities_Init', 'register_routes' ) );
```
The file is loaded earlier, so we can skip it.

gziolo commented on composer.json:
```diff
@@ -1,33 +1,44 @@
 			"phpstan/extension-installer": true,
 			"composer/installers": true
-		}
+		},
```
I moved a few things around, so the order is similar to https://github.com/WordPress/php-ai-client/blob/trunk/composer.json. That made auditing easier for me.

justlevine commented on .gitattributes:
```diff
@@ -0,0 +1,45 @@
+
+# Node and build artifacts.
+node_modules/ export-ignore
+node_modules/** export-ignore
```
Nit: there shouldn't be a need to exclude paths that are already excluded from the repo.

jeffpaul replied:
I think the line above will cover for the second (aka I think we can remove this second line of `/node_modules/** export-ignore`)

justlevine commented on .gitattributes:
```diff
@@ -0,0 +1,45 @@
+/docs/** linguist-documentation
+/CONTRIBUTING.md linguist-documentation
+/LICENSE.md linguist-documentation
+/README.md linguist-documentation
```
1. Are we intentionally including docs?
Usually Composer libraries leave it out of the dist.

2. Either way it should be the same as `package.json:files`, so we should either add `CONTRIBUTING.md` there or also `export-ignore` it above in this file.

gziolo replied:
I thought it would be helpful to include docs, but I'm happy to ignore the folder. We definitely don't want the want about contributing guidelines. Updated with [7fd57a3](https://github.com/WordPress/abilities-api/pull/35/commits/7fd57a3b7f7f9a41490e0771fdc132272e1593ae).

justlevine commented on .gitattributes:
```diff
@@ -0,0 +1,45 @@
+/phpunit.xml.dist export-ignore
+
+# Continuous Integration configs.
+/.github/ export-ignore
```
we should either `export-ignore` `.wordpress-org` too (or just delete the dir entirely, since we're not planning to release this as a plugin so don't need that folder to store assets)

gziolo replied:
I will delete `.wordpress-org` for now.

gziolo replied:
Done in [7fd57a3](https://github.com/WordPress/abilities-api/pull/35/commits/7fd57a3b7f7f9a41490e0771fdc132272e1593ae).

justlevine commented on abilities-api.php:
```diff
@@ -27,26 +27,17 @@
+ * `includes/bootstrap.php` (registered in composer.json) which makes the
+ * procedural functions and classes available. Otherwise fall back to the
+ * local includes bootstrap to keep compatibility with non-Composer installs.
  */
```
I'm thinking we can undo this (and the extra `includes/bootstrap.php`), as we don't need to use Composer's autoloader _internally_ in order for users to require/autoload it in their Composer projects.

(Might even be explicitly better _not_ to autoload and instead have the user `if ( ! class_exists('WP_Ability) ) { require_once 'path/to/vendor/wordpress/abilities-api/abilities-api.php'; }` to avoid conflicts.)

Will test downstream and report back (if nobody beats me to it).

gziolo replied:
Updated with [98d19b0](https://github.com/WordPress/abilities-api/pull/35/commits/98d19b09aed031075697ec2782bd8c751dd8aca5).

jeffpaul commented on .gitattributes:
```diff
@@ -0,0 +1,45 @@
+
+# Tests and test fixtures
+/tests/ export-ignore
+/tests/** export-ignore
```
I think the line above will cover for the second (aka I think we can remove this second line of `/tests/** export-ignore`)

justlevine replied:
if it doesn't, then moving the `.gitkeep export-ignore` higher in the file should do the trick

gziolo replied:
It worked with [7fd57a3](https://github.com/WordPress/abilities-api/pull/35/commits/7fd57a3b7f7f9a41490e0771fdc132272e1593ae) 👍🏻

jeffpaul commented on .gitattributes:
```diff
@@ -0,0 +1,45 @@
+
+# Continuous Integration configs.
+/.github/ export-ignore
+/.github/** export-ignore
```
I think the line above will cover for the second (aka I think we can remove this second line of `/.github/** export-ignore`)

jeffpaul commented on .gitattributes:
```diff
@@ -0,0 +1,45 @@
+
+# Composer/vendor and dev tooling.
+/vendor/ export-ignore
+/vendor/** export-ignore
```
I think the line above will cover for the second (aka I think we can remove this second line of `/vendor/** export-ignore`)

justlevine replied:
I dont think we need to export-ignore anything that's already `.gitignore`d

jeffpaul commented on .gitattributes:
```diff
@@ -0,0 +1,45 @@
+/vendor/** export-ignore
+/composer.lock export-ignore
+
+# Don't include the dotfiles for distribution.
```
I think you can simplify this section to:
`/.* export-ignore`

gziolo replied:
Make sense, updated with [7fd57a3](https://github.com/WordPress/abilities-api/pull/35/commits/7fd57a3b7f7f9a41490e0771fdc132272e1593ae).

jeffpaul commented:
The `composer.json` changes look ok to me, left some comments on the `.gitattributes` file where I think we can simplify it in a couple places.

felixarntz reviewed (commented):
@gziolo Overall this looks solid to me, though I have a high level question about the approach in https://github.com/WordPress/abilities-api/issues/13#issuecomment-3212006337 - why do we want to publish this on Packagist (vs having it as a plugin on .org and thus also on wpackagist.org)?

gziolo commented:
>  why do we want to publish this on Packagist (vs having it as a plugin on .org and thus also on wpackagist.org)?

I responded in the issue. The short version is that, for now, I was willing to replicate how WP Feature API was structured for distribution, and it turns out they offered a Composer library. I updated the config to reflect it in [55fab93](https://github.com/WordPress/abilities-api/pull/35/commits/55fab93e8b06c9a77022f8541b3f8917e5cfe48e).

I addressed all the feedback. This is what I see in the archive fter executing `git archive --format=tar HEAD -o repo.tar && tar -tf repo.tar`:

<img width="368" height="323" alt="Screenshot 2025-08-22 at 11 34 18" src="https://github.com/user-attachments/assets/bac1c953-b460-4f8c-b151-febb51c151fd" />

Should we also exclude the top-level `abilities-api.php`, which is an entry point for the WP plugin?

justlevine commented on composer.json:
```diff
@@ -1,33 +1,44 @@
 	"license": "GPL-2.0-or-later",
-	"type": "wordpress-plugin",
-	"version": "0.1.0",
+	"type": "library",
```
Strongly disagree with this change. Nothing prevents anybody from using a `wordpress-plugin` as a library. But if anybody (including MCP Adapter or Experiments) tries to test this using Composer as a plugin dependency, they can't if it's a `library`.

justlevine replied:
https://github.com/WordPress/abilities-api/pull/42/files#r2294029457

justlevine commented:
> Should we also exclude the top-level abilities-api.php, which is an entry point for the WP plugin?

Please give me time to test using `abilities-api.php` _instead_ of adding `bootstrap.php`, and some of the other changes here that IMO should be necessary for this to be used as a Composer library, don't make sense in our context (versus the WP Feature API prior art), and might even be detrimental to adoption and usage both in the next 3 months and in the post-6.9-merge cleanup.

emdashcodes commented:
> why do we want to publish this on Packagist (vs having it as a plugin on .org and thus also on wpackagist.org)?

> I responded in the issue. The short version is that, for now, I was willing to replicate how WP Feature API was structured for distribution, and it turns out they offered a Composer library. I updated the config to reflect it in [55fab93](https://github.com/WordPress/abilities-api/pull/35/commits/55fab93e8b06c9a77022f8541b3f8917e5cfe48e).

Our main goal for this in the Feature API was to make it possible for plugins to bundle the Feature API as a dependency, without the end user needing to install a separate plugin. This was the reason for the version loader as well.

justlevine commented:
Hey @gziolo et-al please take a look at https://github.com/WordPress/abilities-api/pull/42 which reverts the `"type": "library"` to `type:wordpress-plugin` and includes documentation showing how it can still be including as a Composer library (even if the Feature API considerations may not necessarily apply here).

Jameswlepage commented:
Felix, Jeff, and I had a chat about the wordpress-plugin vs library type discussion. Let's go with the library type to unblock things, mainly so other plugins like Elementor/Woo/Hosting MU etc can include this as a traditional package. This decision is made in part with the understanding that this will go core in 6.9, so we’re treating this as a transitional setup.

We recognize the concern that calling it a library may limit some Composer use cases (e.g., plugin dependencies). However, since wordpress-plugin can still be consumed as a library but not always vice versa, we’ll document usage patterns clearly so consumers understand the intended workflows. If needed, we can revisit the package type once the 6.9 merge has landed and the distribution model stabilizes.

For now, prioritizing unblock and compatibility with large plugin ecosystems outweighs the downside, and the configuration should make adoption easier in the short term while keeping a path open for refinement later.

justlevine commented:
>Let's go with the library type to unblock things, mainly so other plugins like Elementor/Woo/Hosting MU etc can include this as a traditional package. 

This is a false premise. Elementor/Woo/Hosting MU can all use something with a `"type": "wordpress-plugin"` same way they can if it's type "library". 

> However, since wordpress-plugin can still be consumed as a library but not always vice versa, we’ll document usage patterns clearly so consumers understand the intended workflows. 

To me this reads like a justification for _using_ `wordpress-plugin`. If it does everything that a `library` does, why go with the intentionally limiting one that requires additional documentation?

> For now, prioritizing unblock and compatibility with large plugin ecosystems outweighs the downside, and the configuration should make adoption easier in the short term while keeping a path open for refinement later.

This also seems like a false premise. What becomes "unblocked" or "compatible" by using "library" over "wordpress-plugin" , and how does the configuration make adoption easier in the short term?

gziolo commented:
I rebased this branch and applied a couple of small fixes based on the work shared by @justlevine in https://github.com/WordPress/abilities-api/pull/42. Taking into account the decision to go with type `library` in Composer, is there anything missing in this PR before we land it and create a git tag with `v0.1.0` release?

gziolo commented:
> Not strictly necessary for a v0.1.0, but do you want me to backport the relevant doc changes from https://github.com/WordPress/abilities-api/pull/42 into a new PR?

That would be very helpful!
